vm/hash.go

```diff
@@ -133,24 +134,267 @@ func builtinHashInstanceMethods() []*BuiltInMethodObject {
 					// First arg is index
 					// Second arg is assigned value
 					if len(args) != 2 {
-						return newError("Expect 2 arguments. got=%d", len(args))
+						return initErrorObject(ArgumentErrorClass, "Expect 2 arguments. got=%d", len(args))
 					}
 
 					k := args[0]
 					key, ok := k.(*StringObject)
 
 					if !ok {
-						return newError("Expect index argument to be String. got=%T", k)
+						return initErrorObject(TypeErrorClass, "Expect key to be String. got=%v", k.Class().Name)
 					}
 
-					hash := receiver.(*HashObject)
-					hash.Pairs[key.Value] = args[1]
+					h := receiver.(*HashObject)
+					h.Pairs[key.Value] = args[1]
 
 					return args[1]
 				}
 			},
 		},
 		{
+			// Returns empty hash (no key-value pairs)
+			//
+			// ```Ruby
+			// { a: "Hello", b: "World" }.clear # => {}
+			// {}.clear                         # => {}
+			// ```
+			//
+			// @return [Boolean]
+			Name: "clear",
+			Fn: func(receiver Object) builtinMethodBody {
+				return func(t *thread, args []Object, blockFrame *callFrame) Object {
+					if len(args) != 0 {
+						return initErrorObject(ArgumentErrorClass, "Expect 0 argument. got=%d", len(args))
+					}
+
+					return t.vm.initHashObject(make(map[string]Object))
+				}
+			},
+		},
+		{
+			// Loop through keys of the hash with given block frame. It also returns array of
+			// keys in alphabetical order.
+			//
+			// ```Ruby
+			// h = { a: 1, b: "2", c: [1, 2, 3], d: { k: 'v' } }
+			// h.each_key do |k|
+			//   puts k
+			// end
+			// # => a
+			// # => b
+			// # => c
+			// # => d
+			// ```
+			//
+			// @return [Array]
+			Name: "each_key",
+			Fn: func(receiver Object) builtinMethodBody {
+				return func(t *thread, args []Object, blockFrame *callFrame) Object {
+					if len(args) != 0 {
+						return initErrorObject(ArgumentErrorClass, "Expect 0 argument. got=%d", len(args))
+					}
+
+					if blockFrame == nil {
+						t.returnError("Can't yield without a block")
```

Comment:
https://github.com/goby-lang/goby/commit/9a428b14be809d5c7a4e850a4a1ce04b566996f0
We should improve this method to make it receive error type argument instead